dcos_installer/exhibitor_migrate.py

```diff
@@ -0,0 +1,160 @@
+
+| exit code | State                                                                                                                 |
+| --------- | --------------------------------------------------------------------------------------------------------------------- |
+| 0         | Exhibitor has new expected values OR Has successfully triggered the rolling update                                    |
+| 1         | IO Error when trying to connect to exhibitor (expected if run on an agent)                                            |
+| 2         | Error when reading from exhibitor (Any response other than 200)                                                       |
+| 4         | Rolling update in progress                                                                                            |
+| 8         | At least one config value does not have the expected pre migration value, and automatic migration can not take place  |
+| 16        | Attempting to start the rolling update failed due to a non 200 response from exhibitor                                |
+
+
+This script can be run on any of the DC/OS Master nodes by running (as root)
+
+```
+dcos-shell dcos-exhibitor-migrate-perform
+```
+"""  # noqa
+
+ZOOKEEPER_DATA_DIR_START = '/var/lib/zookeeper/snapshot'
+ZOOKEEPER_LOG_DIR_START = '/var/lib/zookeeper/transactions'
+LOG_INDEX_DIR_START = '/var/lib/zookeeper/transactions'
+
+ZOOKEEPER_DATA_DIR_GOAL = '/var/lib/dcos/exhibitor/zookeeper/snapshot'
+ZOOKEEPER_LOG_DIR_GOAL = '/var/lib/dcos/exhibitor/zookeeper/transactions'
+LOG_INDEX_DIR_GOAL = '/var/lib/dcos/exhibitor/zookeeper/transactions'
+
+
+class ValidationError(Exception):
+    def __init__(self, msg, exit_status=8):
+        self.msg = msg
+        self.exit_status = exit_status
+
+
+def get_config_json(response):
+    if response.status_code == 200 and response.headers['Content-Type'].startswith('application/json'):
+        json = response.json()
+        if 'config' not in json:
+            raise ValidationError(".config property not present in exhibitor response. {}".format(response.json()), 2)
+        return json['config']
+    else:
+        raise ValidationError("Error reading configuration from Exhibitor. {} {}"
+                              .format(response.status_code, response.content), 2)
+
+
+def migration_already_complete(config):
+    if config['rollInProgress'] is True:
+        return False
+    else:
+        return (config['zookeeperDataDirectory'] == ZOOKEEPER_DATA_DIR_GOAL and
+                config['zookeeperLogDirectory'] == ZOOKEEPER_LOG_DIR_GOAL and
+                config['logIndexDirectory'] == LOG_INDEX_DIR_GOAL)
+
+
+def assert_can_migrate(config):
+    if config['rollInProgress'] is True:
+        raise ValidationError("Rolling upgrade in progress. {}".format(config['rollStatus']), 4)
+    else:
+        validate_config_key("ZooKeeper Data Directory", ZOOKEEPER_DATA_DIR_START, config['zookeeperDataDirectory'])
```

Comment:
nit: could get rid of this duplication between validate_config / assert_can_migrate, and a lot of the hand-unrolled loops by doing something like the following, not sure if it makes the code actually easier to read / understand though. I'm fine with the code either way.

```
config_keys = [
    'zookeeperDataDirectory',
    'zookeeperLogDirectory',
    'logIndexDirectory']
config_descriptions = [
    'ZooKeeper Data Directory',
    'ZooKeeper Log Directory',
    'Log Index Directory'
]
start_values = [
    '/var/lib/zookeeper/snapshot',
    '/var/lib/zookeeper/transactions',
    '/var/lib/zookeeper/transactions']
goal_values = [
    '/var/lib/dcos/exhibitor/zookeeper/snapshot',
    '/var/lib/dcos/exhibitor/zookeeper/transactions',
    '/var/lib/dcos/exhibitor/zookeeper/transactions']


class ValidationError(Exception):
    def __init__(self, msg, exit_status=8):
        self.msg = msg
        self.exit_status = exit_status


def is_roll_in_progress(config):
    return config['rollInProgress'] is True


def validate_not_rolling(config):
    if is_roll_in_progress(config):
        raise ValidationError("Rolling update in progress. {}".format(config['rollStatus']), 4)


def validate_config_key(pretty_name, expected, actual):
    if actual != expected:
        raise ValidationError("{} does not match expected value. expected: {} actual: {}"
                              .format(pretty_name, expected, actual))


def validate_all_keys(config, values):
    for key, description, expected in zip(config_keys, config_descriptions, values):
        validate_config_key(description, expected, config[key])


def assert_can_migrate(config):
    validate_not_rolling(config)
    validate_all_keys(config, start_values)


def validate_config(config):
    validate_not_rolling(config)
    validate_all_keys(config, goal_values)
```